Changes for getting MySQL metrics in parallel using multiple connections

nipuntalukdar opened the pull request:
I have made changes to pull the MySQL metrics in parallel by using multiple connections. The connection pool size cannot be greater than 64 though. Added a new command line flag mysql.max.connection to set the connection pool size.
Sometimes the exporter takes long time when we collect metrics from a MySQL server with multiple databases and huge number of tables. For example, today I was running the tool on one of our database server and the exporter was taking on average 10 minutes to collect the metrics. With a connection pool of size 10  (so 10 queries in parallel), the scrape time reduced to almost 1 minute 10 seconds on average.

nussjustin commented on collector/exporter.go:
```diff
@@ -165,226 +177,326 @@ func (e *Exporter) scrape(ch chan<- prometheus.Metric) {
+				ch <- prometheus.MustNewConstMetric(scrapeDurationDesc, prometheus.GaugeValue, time.Since(scrapeTime).Seconds(), "connection")
+			}()
+			defer wg.Done()
+			sessionSettingsRows, err := db.Query(sessionSettingsQuery)
```
This doesn't work. This query just changes a setting in the session, which won

nipuntalukdar replied:
Most probably the session settings apply to the entire client object and all the connections in the connection pool. So, we don't need to do that for each and every connection. This is because the db client object hides the details about when new connections are created and when existing connections in the connection pool closed. 

But the session settings should have been applied once when we created the object, rather than doing that for every scrape. I will change that.

nussjustin replied:
The database/sql package and the drivers don't normally track any such changes and thus they will only apply to the current connection. This is ok with only 1 connection, but breaks with multiple connections. 

For these cases you'd normally use transactions since they guarantee that everything is executed on the same connection. But since we are changing a setting that should apply to all connections it should really be set on the pool itself (as explained in #254)

nipuntalukdar commented on collector/exporter.go:
```diff
@@ -23,7 +25,10 @@ const (
 // Metric descriptors.
 var (
-	scrapeDurationDesc = prometheus.NewDesc(
+	db                 *sql.DB
```
I guess there is not much benefit if we make it part of the exporter object. Then we will need to pass the exporter object to all the functions which need access to DB and may need changes in existing functions for that.

nussjustin replied:
Having the DB as a global brings multiple downsides.

1. It's only ever used by the Exporter and depends on fields of the Exporter. Having it as a global hides the relationship of the Exporter and the DB pool (the Exporter basically "owns" the pool).
2. Currently the API allows you to have multiple, independent Exporter instances that collect from different clusters/pools. By depending on a global variable you break the API since all Exporters share one pool. Even worse this is not visible from the API, since you can still create new Exporters by calling New with different DSNs, but  

You can just pass the db field of the Exporter to the Scrape* functions like this

```
ScrapePerfEventsWaits(e.db, ch)
```

This way you don't need to change much and still avoid the global.

nipuntalukdar replied:
Okay. I will make the change. Globals are not good in general.

But I guess, one mysqld_exporter instance can collect from one mysql server only; at least that is what I understood from code. As DATA_SOURCE_NAME env variable is to be defined like user:password@host:port/database. And also the function parseMycnf in mysqld_exporter.go is getting the configs for one DB server only.

nussjustin replied:
From the point of the exporter as application you are right, but the nice thing about keeping the API as is, is that the whole collector logic can be used as a library in other tools and even extended. Also in most cases separating application and logic like this leads to cleaner code overall. And who knows, maybe someday the exporter will be changed to collect from multiple sources and then the collector package doesn't need to change.

SuperQ replied:
The exporter will never collect from multiple sources.  This violates Prometheus best practices.  Exporters are supposed to be 1:1 with their targets.

nipuntalukdar replied:
Yes. I agree.

nussjustin replied:
You're right, but avoiding the global state is still a good idea IMO. The multiple sources were just an example albeit not a good one.

nipuntalukdar commented on collector/exporter.go:
```diff
@@ -142,21 +148,27 @@ func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
-		e.error.Set(1)
-		return
+	var wg sync.WaitGroup
+	if atomic.LoadInt32(&inited) == 0 {
```
Yes. They are similar. Will replace that with sync.Once

nipuntalukdar commented on collector/exporter.go:
```diff
@@ -142,21 +148,27 @@ func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+			}
+			atomic.StoreInt32(&inited, 1)
+			maxCon := e.collect.MaxMySQLConns
+			if maxCon > 16 {
```
The limit comes from a command line parameter. The max limit I wanted to make was 64, but put 16 in the code. That's a miss. We limit the number connections so that the exporter doesn't create performance issues for other MySQL users, by limiting the number of parallel query the exporter can fire (that's equal to max connection pool size). Generally DBA puts a limit on max number of connections that can be used by an user from server side. We are doing that from Client side also, to be nice.

nussjustin replied:
Makes sense. Maybe move the magic number to a named const and explain it in a comment than?

nipuntalukdar replied:
Yes. Will do that.

nipuntalukdar commented on collector/exporter.go:
```diff
@@ -165,226 +177,326 @@ func (e *Exporter) scrape(ch chan<- prometheus.Metric) {
+		wg.Add(1)
+		go func() {
+			wg.Done()
+			scrapeTime = time.Now()
```
Yes. I missed that. scrapeTime should have been a local variable to the go routine function. Making "scrapeTime" a local variable will solve the problem. Don't want synchronization here as that will drastically affect the concurrency we are trying to achieve .

nipuntalukdar commented on mysqld_exporter.go:
```diff
@@ -133,6 +133,10 @@ var (
 	).Default("heartbeat").String()
+	mysqlMaxconns = kingpin.Flag(
+		"mysql.max.connection",
+		"Maximum connection pool size to MySQL server (max value 64)",
```
Correct. I will fix this.

nipuntalukdar commented on collector/exporter.go:
```diff
@@ -165,226 +177,326 @@ func (e *Exporter) scrape(ch chan<- prometheus.Metric) {
-		ch <- prometheus.MustNewConstMetric(scrapeDurationDesc, prometheus.GaugeValue, time.Since(scrapeTime).Seconds(), "collect.global_status")
+		wg.Add(1)
+		go func() {
+			wg.Done()
```
Will correct that.

SuperQ commented:
I'm not sure this is something we can accept.  The difficulty is that it breaks some of the design ideas of the exporter.

Remember that there could be more than one Prometheus or user polling the exporter simultaneously.  This could end up creating many hundreds or thousands of connections pretty quickly.

Each Prometheus scrape is intended to be a single independent probe of the target.  By multi-processing this request, we lose the simplicity of this design.

nussjustin commented:
It would be interesting to know how long each query / scrape function takes. I'd imagine most functions to be pretty fast even with more/bigger databases.

I'm not sure if there's much that can be optimized since the code does mostly just read from the db and sending the values directly to the metric channel, but a quick check doesn't hurt.

@nipuntalukdar Could you try to profile this in your environment? Since the exporter uses the http.DefaultServeMux it should be enough to just import _ "net/http/pprof" in the main package and collect a profile via [pprof](https://blog.golang.org/profiling-go-programs).

nipuntalukdar commented:
@SuperQ the exporter is not fast when the DB server is having many databases with hundreds of tables. When we were trying to use the exporter with few of our database servers, the slowness was the biggest problem and the exporter was taking on average 10 minutes to scrape the metrics. When I changed it to make the queries in parallel, the scrape time was reduced by almost 90%. 

Also, the exporter cannot end up creating many hundreds or thousands of connections, because it is limited by the connection pool size of the DB object the exporter use. Of course we must ensure that there is just one instance of the DB object in the exporter (which I fixed, as previously every scrape was creating and destroying a separate DB object) and generally that is the pattern followed everywhere (just one DB object for one process). If the connection pool size is 10, the exporter will be able to fire only 10 requests in parallel. Also, DBAs generally put a connection limit per user so that one user cannot bog down the database server by abusing it.

Also, will there be 100s of prometheus server scraping the metrics from the same exporter? This doesn't look very practical. But I am new to Prometheus and no idea how people are using it. In such cases the exporter should implement a wait timeout (I will do that if you people agree). Basically, allow first 100s of the requests to use the DB object, the 101 th has to wait till one of the previous 100 completes scraping (and if after timeout also it doesn't get its turns, return with an error). But we can set such wait timeout from the DB object setting itself (connection wait timeout)

nipuntalukdar commented:
@nussjustin 
I think collector_duration_seconds (scrapeDurationDesc) metrics will give those details. But will run and check.

SuperQ commented:
No, we do not globally pool the connection in the exporter.  Each scrape creates a new DB object and connection.  This provides a full end to end test of the MySQL connection in order to satisfy the `mysql_up` metric.

If you have metrics that respond too slowly due to too many tables, those collectors probably need to be disabled for that instance.

nipuntalukdar commented:
@SuperQ I guess using connection pool and reusing the DB object won't affect the mysql_up metric. Only caution we have to take is to ensure every scrape fires the upQuery first and if it detects db server is up, then only subsequent queries in parallel else return from there. 

Also, may be firing the subsequent queries in parallel will help us to get a better point in time state of the mysql metrics. Giving user the freedom to control the concurrency in scraping the metrics (by mysql.max.connection flag) may be helpful. --mysql.max.connection=1 will make the behavior almost like it is currently (with the difference 1. the queries other than upQuery may get fired out of order and 2. even multiple scrapes fired in parallel by different Prometheus servers become sequential due to use of a single DB object with a connection pool size of 1)

SuperQ commented:
Re-using a connection is not a full test of `mysql_up`.  Making a full new connection explicitly part of the test as we want to exercise the new connection and auth path for every scrape request.

SuperQ commented:
This has been implemented in https://github.com/prometheus/mysqld_exporter/pull/265
